File: app/api/routes/learning.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from fastapi import APIRouter, Depends
from pydantic import BaseModel

from app.api.dependencies import get_learning_service
from app.application.learning_service import LearningService


router = APIRouter(prefix="/users/{user_id}", tags=["learning"])


class ProgressItem(BaseModel):
    topic_name: str
    topic_slug: str
    completion_percent: int
    confidence_score: int
    last_studied_at: datetime


class RevisionItem(BaseModel):
    task_id: UUID
    topic_name: str
    topic_slug: str
    due_at: datetime
    reason: str


class SubjectSummary(BaseModel):
    code: str
    name: str
    topics_count: int


class UserDashboardResponse(BaseModel):
    user_id: UUID
    progress: list[ProgressItem]
    due_revisions: list[RevisionItem]
    available_subjects: list[SubjectSummary]


@router.get("/dashboard", response_model=UserDashboardResponse)
async def get_dashboard(
    user_id: UUID,
    learning_service: LearningService = Depends(get_learning_service),
) -> dict[str, Any]:
    return await learning_service.get_user_dashboard(user_id)

File: web/src/types/learning.ts
export interface ProgressItem {
  topic_name: string;
  topic_slug: string;
  completion_percent: number;
  confidence_score: number;
  last_studied_at: string;
}

export interface RevisionItem {
  task_id: string;
  topic_name: string;
  topic_slug: string;
  due_at: string;
  reason: string;
}

export interface SubjectSummary {
  code: string;
  name: string;
  topics_count: number;
}

export interface UserDashboard {
  user_id: string;
  progress: ProgressItem[];
  due_revisions: RevisionItem[];
  available_subjects: SubjectSummary[];
}
